### PR TITLE
Request compression: non-streaming operations

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Adding request compression support for non-streaming operations (e.g., CloudWatch `putMetricData`), excluding S3 operations.
To add support for compression for streaming operations and S3 in future PRs.

## Modifications

Added codegen updates and configuration options.
`CompressRequestStage` handler in `AmazonSyncHttpClient` / `AmazonAsyncHttpClient` to resolve config settings and compress the request if necessary.
`Compressor` / `GzipCompressor` / `CompressorType` for compressing payload.

## Testing

Added unit tests, codegen tests, and CloudWatch `putMetricData` integ tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            input.appendHeader("Content-encoding", compressor.compressorType());
+        } else {
+            input.putHeader("Content-encoding", compressor.compressorType());
+        }
+    }
+
+    private static void updateContentLengthHeader(SdkHttpFullRequest.Builder input) {
+        InputStream inputStream = input.contentStreamProvider().newStream();
+        try {
+            byte[] bytes = IoUtils.toByteArray(inputStream);
```
Comment (L-Applin) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java:116]: I guess there is no other way than buffering the whole request in memory...

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+        if (context.originalRequest().overrideConfiguration().isPresent()
+            && context.originalRequest().overrideConfiguration().get().requestCompressionConfiguration().isPresent()) {
+            Boolean requestCompressionEnabled = context.originalRequest().overrideConfiguration().get()
+                                                       .requestCompressionConfiguration().get()
+                                                       .requestCompressionEnabled();
+            if (requestCompressionEnabled != null) {
+                return requestCompressionEnabled;
+            }
+        }
```
Comment (L-Applin) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java:147]: Nit: suggestion to potentially simplify optional chaining?
```suggestion
        Optional<Boolean> requestCompressionEnabled =
            context.originalRequest().overrideConfiguration()
                   .flatMap(RequestOverrideConfiguration::requestCompressionConfiguration)
                   .map(RequestCompressionConfiguration::requestCompressionEnabled);

        if (requestCompressionEnabled.isPresent()) {
            return requestCompressionEnabled.get();
        }

```
Would something like that work?

**Review by zoewangg — COMMENTED** [review]
Did we update codegen to fail the build if the trait is added to streaming operations

**Comment by davidh44** [comment]
> Did we update codegen to fail the build if the trait is added to streaming operations

Yup, error is thrown in `RequestCompressionTrait`

**Review by L-Applin — APPROVED** [review]

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                   .flatMap(RequestOverrideConfiguration::requestCompressionConfiguration)
+                   .map(RequestCompressionConfiguration::requestCompressionEnabled);
+        if (requestCompressionEnabledRequestLevel.isPresent()) {
+            return requestCompressionEnabledRequestLevel.get();
+        }
+
+        if (compressionEnabledClientLevel != null) {
+            return compressionEnabledClientLevel;
+        }
+        if (context.executionAttributes().getAttribute(SdkExecutionAttribute.REQUEST_COMPRESSION_CONFIGURATION) != null) {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java:159]: As mentioned offline, let's move the resolution logic to SdkDefaultClientBuilder to keep all client options in one place

**Pull request merged**